adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,508 @@
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_048: [ If the tlsio state is TLSIO_APPLEIOS_STATE_OPENING, or TLSIO_APPLEIOS_STATE_CLOSING, the tlsio_appleios_close shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSE, and return _LINE_. ]*/
+            LogError("Try to close the connection with an already closed TLS.");
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+            result = __FAILURE__;
+        }
+        else
+        {
+            CFReadStreamClose(tlsio_instance->sockRead);
+            CFWriteStreamClose(tlsio_instance->sockWrite);
+            /* Codes_SRS_TLSIO_APPLEIOS_32_047: [ If tlsio_appleios_close get success to start the process to close the ssl connection, it shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSING, and return 0. ]*/
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_CLOSING;
+            /* Codes_SRS_TLSIO_APPLEIOS_32_050: [ If the tlsio_appleios_close get success closing the tls connection, it shall call the tlsio_appleios_dowork. ]*/
+            tlsio_appleios_dowork(tlsio_handle);
+            result = 0;
+        }
+    }
+
+    return result;
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_052: [ The tlsio_appleios_send shall send all bytes in a buffer to the ssl connectio. ]*/
+int tlsio_appleios_send_async(CONCRETE_IO_HANDLE tlsio_handle, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+    TLS_IO_INSTANCE* tlsio_instance = (TLS_IO_INSTANCE*)tlsio_handle;
+
+    if ((tlsio_handle == NULL) || (buffer == NULL) || (size == 0))
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_059: [ If the tlsio_handle is NULL, the tlsio_appleios_send shall not do anything, and return _LINE_. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_060: [ If the buffer is NULL, the tlsio_appleios_send shall not do anything, and return _LINE_. ]*/
+        /* Codes_SRS_TLSIO_APPLEIOS_32_061: [ If the size is 0, the tlsio_appleios_send shall not do anything, and return _LINE_. ]*/
+        LogError("Invalid parameter");
+        result = __FAILURE__;
+    }
+    else if (tlsio_instance->state != TLSIO_APPLEIOS_STATE_OPEN)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_058: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, TLSIO_APPLEIOS_STATE_OPENING, TLSIO_APPLEIOS_STATE_CLOSING, or TLSIO_APPLEIOS_STATE_CLOSED, the tlsio_appleios_send shall call the on_send_complete with IO_SEND_ERROR, and return _LINE_. ]*/
+        LogError("TLS is not ready to send data");
+        result = __FAILURE__;
+    }
+    else
+    {
+        CFIndex send_result;
+        size_t send_size = size;
+        const uint8_t* runBuffer = (const uint8_t *)buffer;
+        result = __FAILURE__;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_055: [ if the ssl was not able to send all data in the buffer, the tlsio_appleios_send shall call the ssl again to send the remaining bytes. ]*/
+        while (send_size > 0)
```

Comment:
>while (send_size > 0) [](start = 8, length = 21)

This is not async. To be async, this function should store the information (buffer, callback, etc) and leave the send steps to the dowork.